pkg/rpcserver/rpcserver.go

```diff
@@ -232,17 +234,37 @@ func (serv *server) Serve(ctx context.Context) error {
 	g, ctx := errgroup.WithContext(ctx)
 	g.Go(func() error {
 		return serv.serv.Serve(ctx, func(ctx context.Context, conn *flatrpc.Conn) error {
-			return serv.handleConn(ctx, g, conn)
+			err := serv.handleConn(ctx, conn)
+			if err != nil {
+				log.Logf(0, "serv.handleConn returend %v", err)
+			}
+			return err
 		})
 	})
+	g.Go(func() error {
+		var info *handshakeResult
+		select {
+		case <-ctx.Done():
+			return nil
+		case info = <-serv.onHandshake:
+		}
+		serv.checkOnce.Do(func() {
+			// We run the machine check specifically from the top level context,
+			// not from the per-connection one.
+			g.Go(func() error {
```

Comment:
Do we need double nested goroutines here? Since we already started serv.serv.Serve in a separate goroutine, it looks like we we could do runCheck w/o any goroutines.